File: scripts/embedding.m
function [] = embedding(algo, dataset)
    disp(sprintf("Start with %s and %s...", algo, dataset));
    % execute this function as a script:
    %   matlab -batch "embedding <algo> <dataset>"
    dataset_file = sprintf("../datasets/%s.json", dataset); 
    data = jsondecode(fileread(dataset_file));
    train_triplets = data.train_triplets;
    dim = 2;
    % matlab is 1-indexed 
    train_triplets = train_triplets + 1; 
    
    library = 'vanderMaaten';
    addpath(genpath('../lib/vanderMaaten_STE'));
    startT = cputime;
    switch algo
        case 'CKL'
            mu = 0;  % function doesn't provide a default 
            embedding = ckl_x(train_triplets, dim, mu);
        case 'CKL-K'
            mu = 0;  % function doesn't provide a default
            kernel, embedding = ckl_k(train_triplets, dim, mu);
        case 'GNMDS'
            % mu=[], such that the default is used
            embedding = gnmds_x(train_triplets, [], dim);
        case 'GNMDS-K'
            % mu=[], such that the default is use
            kernel, embedding = gnmds_k(train_triplets, [], dim);
        case 'STE'
            % lambda=[], such that the default is used
            embedding = ste_x(train_triplets, [], dim);
        case 'STE-K'
            % lambda=[], such that the default is used
            kernel, embedding = ste_k(train_triplets, [], dim);
        case 'tSTE'
            embedding = tste(train_triplets, dim);
        otherwise
            error('Unknown algorithm %s', algo)
    end
    endT = cputime;
    
    
    
    result = struct("algorithm", algo, "dataset", dataset, "cpu_time", endT - startT, "library", library, "embedding", embedding);
    if exist("kernel", "var")
        result.kernel = kernel;
    end

    result_file = sprintf("../results/matlab_%s_%s_%s.json", algo, dataset, string(datetime())); 
    disp(sprintf("Write result to %s ...", result_file))
    f = fopen(result_file, 'w');
    fprintf(f, jsonencode(result, PrettyPrint=true));
    fclose(f);
end

File: scripts/embedding.py
import os

from pathlib import Path
import argparse
import json
import time
import numpy as np


DATA_HOME = Path(os.environ.get('CBLEARN_DATA', Path(__file__).parent / '../datasets'))


parser = argparse.ArgumentParser()
parser.add_argument('algo', type=str, help='The embedding algorithm.')
parser.add_argument('dataset', type=str, help='The dataset file.')
#parser.add_argument('result', type=str, help='The result file.')
args = parser.parse_args()
algo = args.algo
dataset =  args.dataset
dims = 2

dataset_file = (DATA_HOME / dataset).with_suffix('.json')
result_file = Path(__file__).parent / f'../results/python_{algo}_{dataset}_{int(time.time())}.json'

with dataset_file.open('r') as f:
    data = json.load(f)

train_triplets = np.array(data['train_triplets'])

### CBLEARN
library = 'cblearn'
from cblearn import embedding
algo_parts = algo.split('-')
is_kernel = 'K' in algo_parts
if 'GPU' in algo_parts:
    backend = 'torch'
    device = 'cuda'
else:
    backend = 'scipy'
    device = 'cpu'

match algo_parts[0]:
    case 'MLDS':
        estimator = embedding.MLDS(dims)  # should raise error for dims != 1
    case 'CKL':
        estimator = embedding.CKL(dims, backend=backend, device=device, kernel=is_kernel)
    case 'FORTE':
        if backend != 'torch':
            raise ValueError(f"FORTE does not support multiple backends at the moment.")
        if is_kernel:
            raise ValueError(f"FORTE does not support kernel at the moment.")
        estimator = embedding.FORTE(dims)
    case 'GNMDS':
        estimator = embedding.GNMDS(dims, backend=backend, device=device, kernel=is_kernel)
    case 'SOE':
        estimator = embedding.SOE(dims, n_init=1, backend=backend, device=device)
    case 'STE':
        estimator = embedding.STE(dims, backend=backend, device=device)
    case 'tSTE':
        estimator = embedding.TSTE(dims, backend=backend, device=device)
    case _:
        raise ValueError(f"Unexpected algorithm {algo}.")

start_time = time.process_time()
estimator.fit(train_triplets)
end_time = time.process_time()
print(end_time - start_time)
embedding = estimator.embedding_
loss = estimator.stress_
### END CBLEARN

result = {'dataset': dataset,
          'library': library,
          'algorithm': algo,
          'loss': float(loss),
          'cpu_time': float(end_time - start_time),
          'embedding': embedding.tolist()}
print(f"Save results to {result_file} ...")
with result_file.open('w') as f:
    json.dump(result, f, indent=4)
